File: tests/block_props_test.cpp
// Copyright (C) by Josh Blum. See LICENSE.txt for licensing information.

#include <boost/test/unit_test.hpp>
#include <iostream>

#include <gras/block.hpp>

struct MyBlock : gras::Block
{
    MyBlock(void):
        gras::Block("MyBlock")
    {
        foo = 0;
        this->register_getter("foo", &MyBlock::get_foo);
        this->register_setter("foo", &MyBlock::set_foo);
    }

    //dummy work
    void work(const InputItems &, const OutputItems &){}

    size_t get_foo(void)
    {
        return foo;
    }

    void set_foo(const size_t &new_foo)
    {
        foo = new_foo;
    }

    size_t foo;
};

BOOST_AUTO_TEST_CASE(test_property_set_get_with_return)
{
    MyBlock my_block;
    BOOST_CHECK_EQUAL(my_block.foo, size_t(0));

    my_block.set("foo", size_t(42));
    BOOST_CHECK_EQUAL(my_block.foo, size_t(42));

    const size_t my_foo = my_block.get<size_t>("foo");
    BOOST_CHECK_EQUAL(my_foo, size_t(42));
}

BOOST_AUTO_TEST_CASE(test_property_set_get_with_reference)
{
    MyBlock my_block;
    BOOST_CHECK_EQUAL(my_block.foo, size_t(0));

    my_block.set("foo", size_t(42));
    BOOST_CHECK_EQUAL(my_block.foo, size_t(42));

    size_t my_foo; my_block.get("foo", my_foo);
    BOOST_CHECK_EQUAL(my_foo, size_t(42));

    double my_foo_d; my_block.get("foo", my_foo_d);
    BOOST_CHECK_EQUAL(my_foo_d, double(42));
}

BOOST_AUTO_TEST_CASE(test_property_errors)
{
    MyBlock my_block;

    //property does not exist
    BOOST_CHECK_THROW(my_block.get<size_t>("bar"), std::exception);

    //wrong type for property
    BOOST_CHECK_THROW(my_block.set("foo", "a string"), std::exception);
}

struct MyBlockNewShit : gras::Block
{
    MyBlockNewShit(void):
        gras::Block("MyBlockNewShit")
    {
        this->register_call("foo", &MyBlockNewShit::foo);
        this->register_call("bar", &MyBlockNewShit::bar);
        this->register_call("baz", &MyBlockNewShit::baz);
        //this->register_call("example_setter", &MyBlockNewShit::example_setter);
    }

    //dummy work
    void work(const InputItems &, const OutputItems &){}

    int foo(void)
    {
        return 42;
    }

    int bar(const std::string &a0)
    {
        return a0.size();
    }

    int baz(const std::string &a0, const bool &a1)
    {
        return a1?a0.size():-1;
    }

    void example_setter(const long &)
    {
        
    }
};

BOOST_AUTO_TEST_CASE(test_property_newshit)
{
    MyBlockNewShit my_block;
    size_t my_foo = my_block.call<size_t>("foo");
    BOOST_CHECK_EQUAL(my_foo, size_t(42));

    size_t my_bar = my_block.call<size_t>("bar", "hello");
    BOOST_CHECK_EQUAL(my_bar, size_t(5));

    signed my_baz = my_block.call<signed>("baz", "hello", false);
    BOOST_CHECK_EQUAL(my_baz, signed(-1));

    //my_block.call<int>("example_setter", 123);
}

File: tests/block_props_test.py
# Copyright (C) by Josh Blum. See LICENSE.txt for licensing information.

import unittest
import gras
import numpy

class MyBlock(gras.Block):
    def __init__(self):
        gras.Block.__init__(self, "MyBlock")
        self.foo = 0
        self.register_property("foo", self.get_foo, self.set_foo)

    def work(self, *args): pass

    def get_foo(self):
        return self.foo

    def set_foo(self, new_foo):
        print "new_foo", new_foo
        new_foo + 0 #throws if its not a number
        self.foo = new_foo

class BlockPropsTest(unittest.TestCase):

    def test_property_set_get(self):
        my_block = MyBlock()
        self.assertEqual(my_block.foo, 0)

        my_block.set("foo", 42)
        self.assertEqual(my_block.foo, 42)

        my_foo = my_block.get("foo")
        self.assertEqual(my_foo, 42)

    def test_property_errors(self):
        my_block = MyBlock()

        #property does not exist
        threw = False
        try: my_block.get("bar")
        except: threw = True
        self.assertTrue(threw)

        #wrong type for property
        threw = False
        try: my_block.set("foo", None)
        except: threw = True
        self.assertTrue(threw)

    def test_element_tree_paths(self):
        my_block = MyBlock()
        tb = gras.TopBlock()
        hb = gras.HierBlock()

        tb.adopt_element("my_hier", hb)
        hb.adopt_element("my_block", my_block)

        my_block.set("foo", 42)
        self.assertEqual(my_block.get("foo"), 42)

        my_block0 = tb.locate_block('/my_hier/my_block')
        self.assertEqual(my_block0.get("foo"), 42)

        my_block1 = hb.locate_block('my_block')
        self.assertEqual(my_block1.get("foo"), 42)

        my_block2 = hb.locate_block('./../my_hier/my_block')
        self.assertEqual(my_block2.get("foo"), 42)

        threw = False
        try: hb.locate_block('../../my_hier/my_block')
        except: threw = True
        self.assertTrue(threw)

        threw = False
        try: hb.locate_block('../../my_hier/my_block0')
        except: threw = True
        self.assertTrue(threw)

        threw = False
        try: hb.locate_block('../../my_hier/my_block/test')
        except: threw = True
        self.assertTrue(threw)

if __name__ == '__main__':
    unittest.main()
